mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -192,41 +293,159 @@ impl TendermintCoin {
+        ))
+    }
+
+    #[inline(always)]
+    fn gas_price(&self) -> f64 { self.gas_price.unwrap_or(DEFAULT_GAS_PRICE) }
+
+    #[allow(unused)]
+    async fn get_latest_block(&self) -> MmResult<GetLatestBlockResponse, TendermintCoinRpcError> {
+        let path = AbciPath::from_str(ABCI_GET_LATEST_BLOCK_PATH).expect("valid path");
+
+        let request = GetLatestBlockRequest {};
+        let request = AbciRequest::new(Some(path), request.encode_to_vec(), None, false);
+
+        let response = self.rpc_client().await?.perform(request).await?;
+
+        Ok(GetLatestBlockResponse::decode(response.response.value.as_slice())?)
+    }
+
+    #[allow(unused)]
+    async fn get_block_by_height(&self, height: i64) -> MmResult<GetBlockByHeightResponse, TendermintCoinRpcError> {
+        let path = AbciPath::from_str(ABCI_GET_BLOCK_BY_HEIGHT_PATH).expect("valid path");
+
+        let request = GetBlockByHeightRequest { height };
+        let request = AbciRequest::new(Some(path), request.encode_to_vec(), None, false);
+
+        let response = self.rpc_client().await?.perform(request).await?;
+
+        Ok(GetBlockByHeightResponse::decode(response.response.value.as_slice())?)
+    }
+
+    // We must simulate the tx on rpc nodes in order to calculate network fee.
+    // Right now cosmos doesn't expose any of gas price and fee informations directly.
+    // Therefore, we can call SimulateRequest or CheckTx(doesn't work with using Abci interface) to get used gas or fee itself.
+    pub(super) fn gen_simulated_tx(
+        &self,
+        account_info: BaseAccount,
+        tx_payload: Any,
+        timeout_height: u64,
+        memo: String,
+    ) -> cosmrs::Result<Vec<u8>> {
+        let fee_amount = Coin {
+            denom: self.denom.clone(),
+            amount: 0_u64.into(),
+        };
+
+        let fee = Fee::from_amount_and_gas(fee_amount, GAS_LIMIT_DEFAULT);
+
+        let signkey = SigningKey::from_bytes(&self.priv_key)?;
+        let tx_body = tx::Body::new(vec![tx_payload], memo, timeout_height as u32);
+        let auth_info = SignerInfo::single_direct(Some(signkey.public_key()), account_info.sequence).auth_info(fee);
+        let sign_doc = SignDoc::new(&tx_body, &auth_info, &self.chain_id, account_info.account_number)?;
+        sign_doc.sign(&signkey)?.to_bytes()
+    }
+
+    #[allow(deprecated)]
+    pub(super) async fn calculate_fee(
+        &self,
+        base_denom: Denom,
+        tx_bytes: Vec<u8>,
+    ) -> MmResult<Fee, TendermintCoinRpcError> {
+        let path = AbciPath::from_str(ABCI_SIMULATE_TX_PATH).expect("valid path");
+        let request = SimulateRequest { tx_bytes, tx: None };
+        let request = AbciRequest::new(Some(path), request.encode_to_vec(), None, false);
+
+        let response = self.rpc_client().await?.perform(request).await?;
+        let response = SimulateResponse::decode(response.response.value.as_slice())?;
+
+        let gas = response.gas_info.expect("Could not simulate tx gas cost");
```

Comment:
I think we shouldn't rely on the honesty of the RPC

Reply:
What do you mean?

Reply:
I mean RPC may return `response.gas_info == None`, and the parser wouldn't reject such response. We should handle such case without panic

Reply:
It definitely can be the case. The third-party RPC server is something we can't be sure of

Reply:
In that case we can't broadcast the transaction as well. We must simulate the tx in order to calculate fees.

Reply:
Yes, and if it returned `gas_info == None`, you could return `TendermintCoinRpcError::InvalidResponse`.

Reply:
Sorry, I forgot this. Done.